python/ray/data/block.py

```diff
@@ -21,6 +21,43 @@
 KeyType = TypeVar("KeyType")
 AggType = TypeVar("AggType")
 
+# A function that extracts a concrete value from a record in a Dataset, used
+# in ``sort(value_fns...)``, ``groupby(value_fn).agg(Agg(value_fn), ...)``.
+# It can either be None (intepreted as the identity function), the name
+# of a Dataset column, or a lambda function that extracts the desired value
+# from the object.
+KeyFn = Union[None, str, Callable[[T], Any]]
+
+
+def _validate_key_fn(ds: "Dataset", key: KeyFn) -> None:
+    """Check the key function is valid on the given dataset."""
+    try:
+        fmt = ds._dataset_format()
+    except ValueError:
+        # Dataset is empty/cleared, validation not possible.
+        return
+    if isinstance(key, str):
+        if fmt == "simple":
+            raise ValueError("String key '{}' requires dataset format to be "
+                             "'arrow' or 'pandas', was '{}'.".format(key, fmt))
+        # Raises KeyError if key is not present in the schema.
+        schema = ds.schema(fetch_if_missing=True)
+        if len(schema.names) > 0 and key not in schema.names:
+            raise ValueError("The column '{}' does not exist in the "
+                             "schema '{}'.".format(key, schema))
+    elif key is None:
+        if fmt != "simple":
+            raise ValueError(
+                "The `None` key '{}' requires dataset format to be "
+                "'simple', was '{}'.".format(key, fmt))
+    elif callable(key):
+        if fmt != "simple":
+            raise ValueError("Callable key '{}' requires dataset format to be "
+                             "'simple', was '{}'.".format(key, fmt))
+    else:
+        raise ValueError("Invalid key type {} ({}).".format(key, type(key)))
```

Comment:
Nit: A function getting an argument of the wrong type should raise a `TypeError`, while an argument of the correct type but whose value is outside of the accepted domain should result in a `ValueError`. Since this is a helper function for validating API function arguments, this is then the wrong-type case, so this should rais a `TypeError`.
```suggestion
        raise TypeError("Invalid key type {} ({}).".format(key, type(key)))
```